File: exts/jphp-gui-ext/src/main/resources/JPHP-INF/sdk/php/gui/event/UXKeyEvent.php
<?php
namespace php\gui\event;

/**
 * Class UXKeyEvent
 * @package php\gui\event
 * @packages gui, javafx
 */
class UXKeyEvent extends UXEvent
{
    /**
     * @var string
     */
    public $character;

    /**
     * @var string
     */
    public $text;

    /**
     * @var string
     */
    public $codeName;

    /**
     * @var bool
     */
    public $altDown;

    /**
     * @var bool
     */
    public $controlDown;

    /**
     * @var bool
     */
    public $shiftDown;

    /**
     * @var bool
     */
    public $metaDown;

    /**
     * @var bool
     */
    public $shortcutDown;

    /**
     * @param UXKeyEvent $parent
     * @param $sender
     */
    public function __construct(UXKeyEvent $parent, $sender)
    {
    }

    /**
     * @param string $accelerator e.g. Control + R
     * @return bool
     */
    public function matches($accelerator)
    {
    }

    /**
     * Left, right, up, down keys (including the keypad arrows)
     * @return bool
     */
    public function isArrowKey()
    {
    }

    /**
     * All Digit keys (including the keypad digits)
     * @return bool
     */
    public function isDigitKey()
    {
    }

    /**
     * Function keys like F1, F2, etc...
     * @return bool
     */
    public function isFunctionKey()
    {
    }

    /**
     * Navigation keys are arrow keys and Page Down, Page Up, Home, End
     * (including keypad keys).
     *
     * @return bool
     */
    public function isNavigationKey()
    {
    }

    /**
     * Keys that could act as a modifier.
     * @return bool
     */
    public function isModifierKey()
    {
    }

    /**
     * All keys with letters.
     * @return bool
     */
    public function isLetterKey()
    {
    }

    /**
     * All keys on the keypad.
     * @return bool
     */
    public function isKeypadKey()
    {
    }

    /**
     * Space, tab and enter.
     * @return bool
     */
    public function isWhitespaceKey()
    {
    }

    /**
     * All multimedia keys (channel up/down, volume control, etc...).
     * @return bool
     */
    public function isMediaKey()
    {
    }

    /**
     * This value is used to indicate that the keyCode is unknown.
     * Key typed events do not have a keyCode value; this value
     * is used instead.
     *
     * @return bool
     */
    public function isUndefinedKey()
    {
    }
}

File: exts/jphp-gui-ext/src/main/java/org/develnext/jphp/ext/javafx/classes/event/UXKeyEvent.java
package org.develnext.jphp.ext.javafx.classes.event;

import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCombination;
import javafx.scene.input.KeyEvent;
import org.develnext.jphp.ext.javafx.JavaFXExtension;
import php.runtime.annotation.Reflection;
import php.runtime.annotation.Reflection.Getter;
import php.runtime.annotation.Reflection.Name;
import php.runtime.annotation.Reflection.Property;
import php.runtime.annotation.Reflection.Signature;
import php.runtime.env.Environment;
import php.runtime.reflection.ClassEntity;

@Name(JavaFXExtension.NS + "event\\UXKeyEvent")
public class UXKeyEvent extends UXEvent {
    interface WrappedInterface {
        @Property String character();
        @Property String text();

        @Property boolean shiftDown();
        @Property boolean controlDown();
        @Property boolean altDown();
        @Property boolean metaDown();
        @Property boolean shortcutDown();
    }

    public UXKeyEvent(Environment env, KeyEvent wrappedObject) {
        super(env, wrappedObject);
    }

    public UXKeyEvent(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public void __construct(KeyEvent event, Object sender) {
        __wrappedObject = new KeyEvent(sender, null, event.getEventType(), event.getCharacter(), event.getText(), event.getCode(), event.isShiftDown(), event.isControlDown(), event.isAltDown(), event.isMetaDown());
    }

    @Override
    public KeyEvent getWrappedObject() {
        return (KeyEvent) super.getWrappedObject();
    }

    @Getter
    public String getCodeName() {
        return getWrappedObject().getCode().getName();
    }

    @Signature
    public boolean isUndefinedKey() {
        return getWrappedObject().getCode() == KeyCode.UNDEFINED;
    }

    @Signature
    public boolean isArrowKey() {
        return getWrappedObject().getCode().isArrowKey();
    }

    @Signature
    public boolean isDigitKey() {
        return getWrappedObject().getCode().isDigitKey();
    }

    @Signature
    public boolean isFunctionKey() {
        return getWrappedObject().getCode().isFunctionKey();
    }

    @Signature
    public boolean isKeypadKey() {
        return getWrappedObject().getCode().isKeypadKey();
    }

    @Signature
    public boolean isLetterKey() {
        return getWrappedObject().getCode().isLetterKey();
    }

    @Signature
    public boolean isMediaKey() {
        return getWrappedObject().getCode().isMediaKey();
    }

    @Signature
    public boolean isModifierKey() {
        return getWrappedObject().getCode().isModifierKey();
    }

    @Signature
    public boolean isNavigationKey() {
        return getWrappedObject().getCode().isNavigationKey();
    }

    @Signature
    public boolean isWhitespaceKey() {
        return getWrappedObject().getCode().isWhitespaceKey();
    }

    @Signature
    public boolean matches(KeyCombination accelerator) {
        if (accelerator == null) {
            return false;
        }

        return accelerator.match(getWrappedObject());
    }

}
